Guid ua binary primitive

zerox1212 opened the pull request:
For review.

This is UA Guid, but using python UUID. Binary protocol is done on the fly.

Can't merge yet because NodeIds need to be refactored to no longer use ua.Guid() class.

Please comment.

oroulet commented:
just had a short look at NodeId class code and it seems very easy to add support for guid. support seems to be missing for encoding currently, need to add one more "if" line

zerox1212 commented:
Should I close the other PR and go this route? I will need to move a few other changes from that branch to this one.

Do you think we need backwards support for ua.Guid class? Or can we just delete it?

oroulet commented:
Delete it

On Wed, Oct 19, 2016, 20:40 Andrew <notifications@github.com> wrote:

> Should I close the other PR and go this route? I will need to move a few
> other changes from that branch to this one.
>
> Do you think we need backwards support for ua.Guid class? Or can we just
> delete it?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeOpcUa/python-opcua/pull/338#issuecomment-254903489>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACcfzkj8lnsNOcHoRSDcDOL82jJlpyqMks5q1mQ2gaJpZM4KbNGO>
> .
>

zerox1212 commented:
Added fixes you suggested, moved miscellaneous changes from https://github.com/FreeOpcUa/python-opcua/pull/336/ to this PR. Updated guid node id. Need help with encoding that you mention.

oroulet commented on tests/tests_unit.py:
```diff
@@ -128,8 +129,8 @@ def test_custom_variant_array(self):
-        g = ua.Guid()
-        sc = ua.StatusCode()
+        # FIXME what tests to add here?
+        pass
```
Well, test what you think is necessary. Create a guid object, convert to binary and back and test it is the same value. For example

oroulet commented on tests/tests_unit.py:
```diff
@@ -150,8 +151,8 @@ def test_nodeid(self):
-        gid = ua.Guid()
-        g = ua.ByteStringNodeId(gid, 0)
+        gid = uuid.uuid4()
+        g = ua.ByteStringNodeId(str(gid), 0)
```
What is that? Create a guid nodeid or put all guid tests in method over

zerox1212 replied:
I need help here. I do not understand what this test did in the first place...

zerox1212 commented:
Almost ready, but seems python 2.7 doesn't support the very nice int method `to_bytes()` .

oroulet commented:
Yes it looks good!

On Wed, Oct 19, 2016, 21:57 Andrew <notifications@github.com> wrote:

> Almost ready, but seems python 2.7 doesn't support the very nice int
> method to_bytes() .
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeOpcUa/python-opcua/pull/338#issuecomment-254922816>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACcfzoh_aGYw4iuFgl8XuFu6ZdUrRJbVks5q1nYNgaJpZM4KbNGO>
> .
>

oroulet commented:
Looks broken. Write a test you think is good

On Wed, Oct 19, 2016, 21:58 Andrew <notifications@github.com> wrote:

> *@zerox1212* commented on this pull request.
> ------------------------------
>
> In tests/tests_unit.py
> <https://github.com/FreeOpcUa/python-opcua/pull/338>:
>
> > @@ -150,8 +151,8 @@ def test_nodeid(self):
>          s = ua.StringNodeId(53, 0)  # should we raise an exception???
>          s1 = ua.StringNodeId("53", 0)
>          bs = ua.ByteStringNodeId(b"53", 0)
> -        gid = ua.Guid()
> -        g = ua.ByteStringNodeId(gid, 0)
> +        gid = uuid.uuid4()
> +        g = ua.ByteStringNodeId(str(gid), 0)
>
> I need help here. I do not understand what this test did in the first
> place...
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeOpcUa/python-opcua/pull/338>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACcfzleZeKE48ohn2xapZe6SkaqBXS94ks5q1nZOgaJpZM4KbNGO>
> .
>

zerox1212 commented:
I think it is good enough for merge. NodeId tests are the same as before. I'm not sure how to improve them.

zerox1212 commented:
Actually don't merge just yet, I have to test xml exporter.

zerox1212 commented:
Please review, should be ready for merge (I don't know why pypy failed...).

This does bring up a question though:
What to do with instantiate? If an the object that is being instantiated should we automatically create a new UUID? Perhaps we need to add that before merging this.

oroulet commented:
We can fix instanciate later. Maybe there is a bug in instantiate where we
change type of nodeid. But this is not related

On Wed, Oct 19, 2016, 23:38 Andrew <notifications@github.com> wrote:

> Please review, should be ready for merge (I don't know why pypy failed...).
>
> This does bring up a question though:
> What to do with instantiate? If an the object that is being instantiated
> should we automatically create a new UUID? Perhaps we need to add that
> before merging this.
>
> —
> You are receiving this because you commented.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeOpcUa/python-opcua/pull/338#issuecomment-254948595>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACcfzq-fAsopyrReteRukB0tenyy9dL0ks5q1o3MgaJpZM4KbNGO>
> .
>

zerox1212 commented on opcua/common/ua_utils.py:
```diff
@@ -87,13 +87,21 @@ def string_to_val(string, vtype):
+        if string is None or string == '':
+            val = uuid.uuid4()
+        else:
+            val = uuid.UUID(string)
```
Just want to point this out and make sure it's acceptable. It might be better to throw an exception, but in modeler it's a pain to type in a valid Guid. Same for datetime. However this function might be used by something else and have unintended affects.

oroulet commented:
No it is not since null nodeid might be desired. But we could add code in
uamodeller to generate default value

On Thu, Oct 20, 2016, 16:09 Andrew <notifications@github.com> wrote:

*@zerox1212* commented on this pull request.
------------------------------

In opcua/common/ua_utils.py
<https://github.com/FreeOpcUa/python-opcua/pull/338#pullrequestreview-5076612>
:

>      elif vtype == ua.VariantType.LocalizedText:
         val = ua.LocalizedText(string)
     elif vtype == ua.VariantType.StatusCode:
         val = ua.StatusCode(string)
     elif vtype == ua.VariantType.Guid:
-        val = uuid.UUID(string)
+        # if no guid string is supplied, make a random guid (this is
required for modeler)
+        if string is None or string == '':
+            val = uuid.uuid4()
+        else:
+            val = uuid.UUID(string)

Just want to point this out and make sure it's acceptable. It might be
better to throw an exception, but in modeler it's a pain to type in a valid
Guid. Same for datetime. However this function might be used by something
else and have unintended affects.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/FreeOpcUa/python-opcua/pull/338#pullrequestreview-5076612>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ACcfziNsUD0PwuW8WfCHv7PCyDgNrO36ks5q13YcgaJpZM4KbNGO>
.

zerox1212 commented:
I think it's finally ready.

One thing about modeler, I think it would be better to add a new function to python-opcua utils called `get_default` or something. Then modeler can just call `utils.get_default(type)` if the value field is null. This would match the current code organization much better (modeler is a shell, bulk of code in server). What do you think?
